### PR TITLE
[SPIR-V][DOC] Add SPV_INTEL_optimization_hints extension spec

### PR DESCRIPTION
**Pull request by MrSidims**
This extension enables analogs of llvm.assume and llvm.expect
intrinsics in SPIR-V.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+This extension is written against the SPIR-V Specification,
+Version 1.5 Revision 2.
+
+This extension requires SPIR-V 1.0.
+
+Overview
+--------
+
+This extension adds a decoration *AssumeTrueINTEL* and an instruction
```
Comment (AlexeySachkov) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:62]: `assume` is not a decoration anymore

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+----
+ExpectINTEL
+----
+
+New Decorations
+---------------
+Decorations added under the *OptimizationHintsINTEL* capability:
+
+----
+AssumeTrueINTEL
```
Comment (AlexeySachkov) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:98]: This is not a decoration anymore

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+Version
+-------
+
+[width="40%",cols="25,25"]
+|========================================
+| Last Modified Date | 2020-04-27
```
Comment (AlexeySotkin) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:47]: Please keep it up to date

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+This extension is written against the SPIR-V Specification,
+Version 1.5 Revision 2.
+
+This extension requires SPIR-V 1.0.
+
+Overview
+--------
+
+This extension adds *AssumeTrueINTEL* and *OpExpectINTEL* instructions that are
```
Comment (AlexeySotkin) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:62]:
```suggestion
This extension adds *OpAssumeTrueINTEL* and *OpExpectINTEL* instructions that are
```

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+appropriately mapped on ‘llvm.assume’ and ‘llvm.expect’ intrinsics. That is
```
Comment (AlexeySotkin) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:63]:
```suggestion
This extension adds *OpAssumeTrueINTEL* and *OpExpectINTEL* instructions which are
mapped to ‘llvm.assume’ and ‘llvm.expect’ intrinsics respectively. That is
```

Comment (AlexeySotkin) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:63]: I'm not sure about "to"

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+OptimizationHintsINTEL
+----
+
+New Instructions
+----------------
+Instructions added under the *OptimizationHintsINTEL* capability:
+
+----
+ExpectINTEL
+AssumeTrueINTEL
```
Comment (AlexeySotkin) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:90]:
```suggestion
OpExpectINTEL
OpAssumeTrueINTEL
```

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+----
+
+Token Number Assignments
+------------------------
+[width="45%",cols="30,15"]
+|===============================
+| OptimizationHintsINTEL | 5629
+| AssumeTrueINTEL        | 5630
```
Comment (AlexeySotkin) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:98]:
```suggestion
| OpAssumeTrueINTEL        | 5630
```

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+~~~~~~~~~~~~
+
+Modify Section 3.31, "Capability", adding these rows to the Capability table:
+
+--
+[options="header"]
+|====
+2+^| Capability ^| Depends On
+| 5629 | *OptimizationHintsINTEL* +
+Allow to use AssumeTrueINTEL and ExpectINTEL instructions |
```
Comment (AlexeySotkin) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:115]:
```suggestion
Allow to use *OpAssumeTrueINTEL* and *OpExpectINTEL* instructions |
```

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+Instructions
+~~~~~~~~~~~~
+
+In section 3.32.1, Miscellaneous Instructions, add a new instruction:
+
+[cols="3", width="100%"]
+|=====
+2+^|*OpAssumeTrueINTEL* +
+The instruction allows the optimizer to assume that the provided _Condition_ is
+always true whenever the control flow reaches the instruction call. If the
```
Comment (AlexeySotkin) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:128]:
```suggestion
always true whenever the control flow reaches the instruction. If the
```

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+_Condition_ is violated during execution, the behavior is undefined.
+
+The type of _Condition_ shall be _Boolean_.
```
Comment (AlexeySotkin) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:131]:
```suggestion
The type of _Condition_ must be _Boolean_.
```

Comment (MrSidims) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:131]: I disagree. Usually in specification/requirement docs `shall` has a power of `must`, so it's a strong requirement.

Comment (AlexeySotkin) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:131]: Current version of the [SPIRV spec](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html) contains:
 * "shall" 0 times
 * "must" 1390 times

Comment (MrSidims) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc]: Applied. If you don't mind I would leave the revision version unchanged :)

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+In section 3.32.1, Miscellaneous Instructions, add a new instruction:
+
+[cols="3", width="100%"]
+|=====
+2+^|*OpAssumeTrueINTEL* +
+The instruction allows the optimizer to assume that the provided _Condition_ is
+always true whenever the control flow reaches the instruction. If the
+_Condition_ is violated during execution, the behavior is undefined.
+
+The type of _Condition_ must be _Boolean_.
```
Comment (bashbaug) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:131]: Must _Condition_ be a scalar, or may it be a vector of _Boolean type_ also?

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+| 2 | 5630 | Condition <id>
+|=====
+
+[cols="6", width="100%"]
+|=====
+5+^|*OpExpectINTEL* +
+
+Provides information for optimizers that the most probable value of _Value_ is
+_ExpectedValue_. This instruction can be safely replaced with an entity
+referenced by _Value_.
```
Comment (bashbaug) [inline sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:144]: I think we need to strengthen this description somewhat.  Is the intent that **OpExpectINTEL** behaves the same as **OpCopyObject** except that it also provides information about the most probable value of _Value_?

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc (hunk 11 above)
Comment (MrSidims) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:131]: It shall be scalar. Added.

File: sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc (hunk 12 above)
Comment (MrSidims) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc:144]: > Is the intent that OpExpectINTEL behaves the same as OpCopyObject except that it also provides information about the most probable value of Value?

Yes, but OpExpectINTEL is more strict about result type.

Comment (bashbaug) [reply sycl/doc/extensions/SPIRV/SPV_INTEL_optimization_hints.asciidoc]: I still don't think this is sufficiently descriptive, and there is a typo now (`entityi`).

How about something like:

> This instruction behaves the same as *OpCopyObject* by making a copy of _Value_,
except it also provides information to the optimizer that the most probable
value of _Value_ is _ExpectedValue_.
>
> _Result Type_ must be a scalar or vector of _integer type_ or _Boolean type_.
> 
> The type of _Value_ and _ExpectedValue_ must be the same as _Result Type_.

**Comment by MrSidims** [comment]
Closing this PR

**Comment by bader** [comment]
> Closing this PR

Why?

**Comment by MrSidims** [comment]
Sorry, I indeed should specify, why I'm closing the PR. Reason is that this extension is being replaced by KHR one.

**Pull request closed without merging**